### ISSUE TITLE
[mobx-undecorate] need option to treat parse ts file as non-jsx

### ISSUE THREAD

**Issue by xiluo58** [mobxjs/mobx#2754]
One example with error
```
import { useSearch } from "./useSearch";

export function useAlias(): string {
    const parsedSearch = useSearch();
    return (<string>parsedSearch.alias || "").toUpperCase();
}
```
Error message: `Transformation error (Unterminated JSX contents (5:20))
SyntaxError: Unterminated JSX contents (5:20)
`

There is no error if type casting is removed
```
import { useSearch } from "./useSearch";

export function useAlias(): string {
    const parsedSearch = useSearch();
    return (parsedSearch.alias || "").toUpperCase();
}
```

In my project jsx code is only in files with tsx extension. There should be an option to let the parser don't parse js/ts files as jsx

**Comment by FredyC** [comment]
Can you please set up simple repo where we can run this and see the problem? Or ideally, submit a PR with failing tests.

**Comment by iChenLei** [comment]
#### You can use `as`, `<Bracket syntax>` type casting  is conflict with JSX
```diff
import { useSearch } from "./useSearch";

export function useAlias(): string {
    const parsedSearch = useSearch();
-    return (parsedSearch.alias as string || "").toUpperCase();
+    return (<string>parsedSearch.alias as string || "").toUpperCase();
}
```
#### cause it's code refactor, Provide filename extensions option will be good for [mobx-undecorate] user
// source code → packages/mobx-undecorate/cli.js
```js
 #!/usr/bin/env node
const { spawn } = require("child_process")
const path = require("path")

spawn(
    path.join("node_modules", ".bin", "jscodeshift"),
    [
        "--extensions=js,jsx,ts,tsx",
        ...process.argv.filter(arg => arg.startsWith("--")),
        "-t",
        "src/undecorate.ts",
        process.cwd()
    ],
    {
        cwd: __dirname,
        stdio: "inherit",
        shell: true
    }
)
```
### emergent measures for you
open node_modules folder, change the cli code, and run
```diff
-        "--extensions=js,jsx,ts,tsx",
+        "--extensions=jsx,tsx",
```
> btw, `as` syntax is better than `<Bracket>`, especially for react/jsx developer

**Comment by mweststrate** [comment]
We don't support using separate parsers per file type at the moment. So the simplest work around is to upgrade your code to use modern casts instead? I can imagine that some other codemod exists to automate that :)

**Comment by xiluo58** [comment]
The project used https://github.com/microsoft/tslint-microsoft-contrib with prefer-type-cast enabled before. There is a lot of code uses <> to cast type. I haven't found a codemod for it. Even if there is one, make big change is still risky.

What about this?
try {
  // parse as jsx
} catch (e) {
  if (isParseError(e)) {
    // parse as js
  }
}
So the logic is same for all files

**Comment by xiluo58** [comment]
> Can you please set up simple repo where we can run this and see the problem? Or ideally, submit a PR with failing tests.

Here is the repo https://github.com/xiluo58/mobx-undecorate-reproduce/tree/main/type-casting

**Comment by iChenLei** [comment]
@xiluo58 Hi, It should be fixed by #2785 , hope we can release a new version as soon as possible to support parse ts as non-jsx.

**Comment by iChenLei** [comment]
@xiluo58 Hi, we have released mobx-undecorate v1.0.7, now you can use `--parseTsAsNonJsx` to support parse ts file as non-jsx!
https://yarnpkg.com/package/mobx-undecorate
https://www.npmjs.com/package/mobx-undecorate
